internal/dump/common.go

```diff
@@ -11,63 +11,84 @@ import (
+// A Dumper dumps files from a repository.
+type Dumper struct {
+	cache  *bloblru.Cache
+	format string
+	repo   restic.Repository
+	w      io.Writer
 }
 
-// WriteDump will write the contents of the given tree to the given destination.
-// It will loop over all nodes in the tree and dump them recursively.
-type WriteDump func(ctx context.Context, repo restic.Repository, tree *restic.Tree, rootPath string, dst io.Writer) error
+func New(format string, repo restic.Repository, w io.Writer) *Dumper {
+	return &Dumper{
+		cache:  bloblru.New(64 << 20),
+		format: format,
+		repo:   repo,
+		w:      w,
+	}
+}
 
-func NewCache() *bloblru.Cache {
-	return bloblru.New(64 << 20)
+func (d *Dumper) DumpTree(ctx context.Context, tree *restic.Tree, rootPath string) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// ch is buffered to deal with variable download/write speeds.
+	ch := make(chan *restic.Node, 10)
+	go sendTrees(ctx, d.repo, tree, rootPath, ch)
+
+	switch d.format {
+	case "tar":
+		return d.dumpTar(ctx, ch)
+	case "zip":
+		return d.dumpZip(ctx, ch)
+	default:
+		panic("unknown dump format")
+	}
 }
 
-func writeDump(ctx context.Context, repo restic.Repository, tree *restic.Tree, rootPath string, dmp dumper) error {
-	for _, rootNode := range tree.Nodes {
-		rootNode.Path = rootPath
-		err := dumpTree(ctx, repo, rootNode, rootPath, dmp)
-		if err != nil {
-			// ignore subsequent errors
-			_ = dmp.Close()
+func sendTrees(ctx context.Context, repo restic.Repository, tree *restic.Tree, rootPath string, ch chan *restic.Node) {
+	defer close(ch)
 
-			return err
+	for _, root := range tree.Nodes {
+		root.Path = path.Join(rootPath, root.Name)
+		if sendNodes(ctx, repo, root, ch) != nil {
+			break
 		}
 	}
-
-	return dmp.Close()
 }
 
-func dumpTree(ctx context.Context, repo restic.Repository, rootNode *restic.Node, rootPath string, dmp dumper) error {
-	rootNode.Path = path.Join(rootNode.Path, rootNode.Name)
-	rootPath = rootNode.Path
-
-	if err := dmp.dumpNode(ctx, rootNode, repo); err != nil {
-		return err
+func sendNodes(ctx context.Context, repo restic.Repository, root *restic.Node, ch chan *restic.Node) error {
+	select {
+	case ch <- root:
+	case <-ctx.Done():
+		return ctx.Err()
 	}
 
 	// If this is no directory we are finished
-	if !IsDir(rootNode) {
+	if !IsDir(root) {
 		return nil
 	}
 
-	err := walker.Walk(ctx, repo, *rootNode.Subtree, nil, func(_ restic.ID, nodepath string, node *restic.Node, err error) (bool, error) {
+	err := walker.Walk(ctx, repo, *root.Subtree, nil, func(_ restic.ID, nodepath string, node *restic.Node, err error) (bool, error) {
 		if err != nil {
 			return false, err
 		}
 		if node == nil {
 			return false, nil
 		}
 
-		node.Path = path.Join(rootPath, nodepath)
+		node.Path = path.Join(root.Path, nodepath)
 
-		if IsFile(node) || IsLink(node) || IsDir(node) {
-			err := dmp.dumpNode(ctx, node, repo)
-			if err != nil {
-				return false, err
-			}
+		switch node.Type {
```

Comment:
I'd prefer to keep the node type checks using `IsFile` etc. instead of checking the node type differently in different places.

Reply:
Done.